### ISSUE TITLE
[FEATURE]Exactly-once sink using transaction

### ISSUE THREAD

**Issue by reswqa** [streamnative/pulsar-flink#165]
**Is your feature request related to a problem? Please describe.**
Our current implementation of pulsar-flink sink can only guarantee the At least-once semantic, but cannot be Exactly-once. This is a serious problem for End-to-End Exactly-once streaming systems like apache flink. With Pulsar's continuous improvement of transaction support, it is time to solve this problem.

**Describe the solution you'd like**
Implement Exactly-once sink based on Pulsar's transaction API

**Describe alternatives you've considered**
Now that we have support for transaction, we don't need to consider the realization of idempotent sink.

**Additional context**
[PIP-31: Transactional Streaming](https://docs.google.com/document/d/145VYp09JKTw9jAT-7yNyFU255FptB2_B2Fye100ZXDI/edit#heading=h.z0ro8zbxdcc3)
[[FLINK-7210] Introduce TwoPhaseCommitSinkFunction](https://github.com/apache/flink/pull/4368)
[Kafka's implement](https://github.com/apache/flink/tree/master/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka)

**Comment by reswqa** [comment]
Some preliminary plans are listed below, and detailed plans will be completed as the work progresses.
1. Investigate the implementation of Kafka connector, sum up what transaction support our implementation requires, and synchronize related information with @gaoran10 .
2. Implement transaction-based sink.
3. Expand to Flink Unified Sink API.

**Comment by reswqa** [comment]
Flink uses the two-phase commit protocol to implement TwoPhaseCommitSinkFunction. The main life cycle methods are beginTransaction(), preCommit(), commit(), abort(), recoverAndCommit(), recoverAndAbort().
Take the implementation of Kafka as an example. For Kafka things, there are three main parameters: transactionId, producerId, and epoch. These three together determine a session of transaction. Therefore, Flink mainly persistently stores the State object composed of these three.
The transactionId is a string configured by the user. When the producer uses the transactionId to initTransaction, the TC will assign the ProducerId and epoch to it, and ensure that the same PID must be assigned to the same TID.

beginTransaction:
Flink maintains a TID pool. Every time a checkPoint comes, it takes out a TID to start a new producer, and obtains its PID and epoch from the TC.

preCommit:
In the preCommit phase, all data is forced to Flush.

commit&abort:
Commit and abort directly call the commit and abort interfaces of Kafka.

recoverAndCommit:
When a Flink job fails, if preCommit has succeeded, we must ensure that transaction is finally committed. We recover the TID, PID, and epoch from the Transaction state, create a Producer and set the (TID,PID,epoch) to the State, and then Commit it immediately. (Can't send the initTransaction message to TC, this will trigger the fence mechanism and cause transaction to abort)

recoverAndAbort:
When the Flink job fails, if the previous transaction has not been preCommited, we must ensure that the transaction is eventually Abort. We restore the TID from the Transaction state, create a Producer and set the PID for it, and perform the initTransaction operation. This will make the TC observe the existence of multiple producers with the same TID, and it will Abort the previous tansaction and allocate a new epoch.

To sum up: For Pulsar's Sink, we have several information that must be clear:
1. What information is used as a persistent State to uniquely identify an ongoing transaction.
2. Is the life cycle of transation on the TC side closer to that of Kafka?
3. When the Flink job is recovering, how the producer side should perform state recovery and corresponding operations for recoverAndCommit and recoverAndAbort respectively.
4. How to control the number of transactions that are opened at the same time? In the implementation of Kafka, a fixed-size Pool is used to recycle the TID. If there is no free PID available, an exception is thrown.

**Comment by gaoran10** [comment]
### How Pulsar flink use Pulsar transaction.

1. client init

```
PulsarClient pulsarClient = PulsarClient.builder()
        .serviceUrl(getPulsarServiceList().get(0).getBrokerServiceUrl())
        .statsInterval(0, TimeUnit.SECONDS)
        .enableTransaction(true)
        .build();
```

2. begin transaction

```
Transaction txn = ((PulsarClientImpl) pulsarClient)
        .newTransaction()
        .withTransactionTimeout(2, TimeUnit.SECONDS)
        .build()
        .get();
```

3. send transaction messages

```
CompletableFuture<MessageId> sendFuture1 = 
		producer.newMessage(txn).value(("Hello Txn").getBytes()).sendAsync();
CompletableFuture<MessageId> sendFuture2 = 
		producer.newMessage(txn).value(("Hello Txn").getBytes()).sendAsync();
```

4. ack transaction messages

```
CompletableFuture<Void> ackFuture = consumer.acknowledgeAsync(message.getMessageId(), txn);
```

5. preCommit

```
guarantee all send and ack futures are fininshed
```

6. commit or abort

```
// normal commit
txn.commit().get();

// recover and commit
((PulsarClientImpl) pulsarClient).getTcClient().commitAsync(
        new TxnID(txnidMostBits, txnidLeastBits),
        Lists.list(sinkMessageId1, sinkMessageId2)).get();
```

**Comment by gaoran10** [comment]
### Demo

```
@Test
public void test() throws Exception {
    String sourceTopic = "txn/ns1/consume-topic";
    String sinkTopic = "txn/ns1/sink-topic";

    ConsumerImpl<byte[]> sourceConsumer = (ConsumerImpl<byte[]>) pulsarClient.newConsumer()
            .topic(sourceTopic)
            .subscriptionName("test")
            .subscribe();

    Producer<byte[]> sourceProducer = pulsarClient.newProducer()
            .topic(sourceTopic)
            .enableBatching(false)
            .create();

    Consumer<byte[]> sinkConsumer = pulsarClient.newConsumer()
            .topic(sinkTopic)
            .subscriptionName("test")
            .subscribe();

    ProducerImpl<byte[]> sinkProducer = (ProducerImpl<byte[]>) pulsarClient.newProducer()
            .topic(sinkTopic)
            .sendTimeout(0, TimeUnit.SECONDS)
            .create();

    for (int i = 0; i < 5; i++) {
        sourceProducer.newMessage().value("source-content".getBytes()).sendAsync();
    }

    TransactionImpl txn = (TransactionImpl) ((PulsarClientImpl) pulsarClient).newTransaction()
            .withTransactionTimeout(30, TimeUnit.MINUTES)
            .build().get();

    List<CompletableFuture<Void>> ackFutureList = new ArrayList<>();
    for (int i = 0; i < 5; i++) {
        Message<byte[]> message = sourceConsumer.receive();
        log.info("source message id {}", message.getMessageId());
        CompletableFuture<Void> ackFuture = sourceConsumer.acknowledgeAsync(message.getMessageId(), txn);
        ackFutureList.add(ackFuture);
    }
    CompletableFuture<MessageId> sinkFuture1 =
            sinkProducer.newMessage(txn).value("sink-content-1".getBytes()).sendAsync();
    CompletableFuture<MessageId> sinkFuture2 =
            sinkProducer.newMessage(txn).value("sink-content-2".getBytes()).sendAsync();

    // save txnidMostBits, txnidLeastBits, and all sink messages ids to checkpoint
    long txnidMostBits = txn.getTxnIdMostBits();
    long txnidLeastBits = txn.getTxnIdLeastBits();
    MessageIdImpl sinkMessageId1 = (MessageIdImpl) sinkFuture1.get();
    MessageIdImpl sinkMessageId2 = (MessageIdImpl) sinkFuture2.get();

    Message<byte[]> sinkMessage = sinkConsumer.receive(5, TimeUnit.SECONDS);
    Assert.assertNull(sinkMessage);

    // normal commit
//        txn.commit().get();
//
//        sinkMessage = consumer.receive(5, TimeUnit.SECONDS);
//        Assert.assertNotNull(sinkMessage);
//        sinkMessage = consumer.receive(5, TimeUnit.SECONDS);
//        Assert.assertNotNull(sinkMessage);

    // preCommit
    // The ack future list and sink future list should be all finished, this is useful for recover operation.
    // recover and commit
    ((PulsarClientImpl) pulsarClient).getTcClient().commitAsync(
            new TxnID(txnidMostBits, txnidLeastBits),
            Lists.list(sinkMessageId1, sinkMessageId2)).get();

    sinkMessage = sinkConsumer.receive(5, TimeUnit.SECONDS);
    Assert.assertNotNull(sinkMessage);
    sinkMessage = sinkConsumer.receive(5, TimeUnit.SECONDS);
    Assert.assertNotNull(sinkMessage);
}
```
